user/man/evltc.1
.\" This manpage has been automatically generated by docbook2man
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "EVLTC" "1" "6 December 2002" "" ""
.SH NAME
evltc \- Compile formatting templates
.SH SYNOPSIS
.fi
\fBevltc\fR \fB  \fIsourcefile\fB  [\fB-f | --func]  [-n | --noto]   
.nf
.fi
        [-c | --cpp]  [cpp_options] \fR

or

\fBevltc\fR \fB \fIbinfile.to\fR
.nf
.SH "DESCRIPTION"
.PP
The \fBevltc\fR command reads the
formatting template specification(s) in
\fIsourcefile\fR and creates a binary
template file for each specification.
\fIsourcefile\fR may not define two templates with
the same \fIevent_type\fR, even if they are for  
different facilities.
Binary files are created
in the directory where \fIsourcefile\fR
resides.
.PP
If the name of the specified file ends in \fI.to\fR, it is assumed to be a
binary template file.  It is read, and if it contains an event-record
template, the corresponding sample call to the \fBevl_log_write()\fR function
is printed, as with \fB-f\fR.  In this case, the \fB-f\fR and \fB-n\fR options (if
specified) are ignored, and any cpp-related options are flagged as
errors.
.PP
The binary file for an event-record template is named
\fIeventtype.to\fR,
where
\fIeventtype\fR is the hexadecimal
event type.
For a facility's default template, the binary file is named
\fIdefault.to\fR.
.PP
The binary file for a struct template is named
\fIstructname.to\fR.
.PP
If the source file contains any errors, no binary files
are produced, and error messages are written to stderr.
.PP
The algorithm for finding a struct template that is referenced by
another template is as follows:
.fi

(1) The struct_path from the import statement is converted into a relative pathname by
replacing all periods with slashes and appending .to.  For example,
gui.graphics.point becomes gui/graphics/point.to. 

(2) This relative pathname is applied to each of the following directories in turn until a
file is found: 
 (a) the directory in which the current template source file resides. 
 (b) the directories specified in the EVLTMPLPATH environment variable (a
colon-delimited list of directories).  If the EVLTMPLPATH environment
variable is not defined, the directories
/var/evlog/templates/$LANG and /var/evlog/templates
are searched, in that order. 
 
(3) The selected template file is read into memory if it is not already there. 

It is neither necessary nor permitted to import a template with the same name
as one previously defined in the current template source file, unless the
imported template is from a different directory.  See Example 3 (below).
.nf
.SH "OPTIONS"
.TP
\fB-f, --func\fR
When this option is specified, a sample call to the
evl_log_write() function is created.
Such a call could be used to log an event record whose 
format matches the template.  
.TP
\fB-n, --noto\fR
Compile the template source and report errors as appropriate, but
do not write out the binary template file(s).
.TP
\fB-c, --cpp\fR
Run \fIsourcefile\fR through the C preprocessor, /lib/cpp, and
compile the result.  The contents of \fIsourcefile\fR are unchanged.
.TP
\fB-D\fR  \fImacro[=defn]\fR
Passed to the C preprocessor.  Implies -c.
.TP
\fB-U\fR  \fImacro\fR
Passed to the C preprocessor.  Implies -c.
.TP              
\fB-I\fR  \fIdir\fR
Passed to the C preprocessor.  Implies -c.
.SH "EXAMPLES"

\fBExample 1.\fR
.nf

Suppose the file \fIVolMgr.t\fR contains the following text.

.nf
/* Start of first template */
struct sourceInfo;
attributes {
      string sourceFile;  int lineNumber;
}
format
%sourceFile%: %lineNumber%:
END

/* Start of second template */
facility "VolMgr"; event_type 0x456;
attributes {
      struct sourceInfo srcInfo;
      string volumeName;  ulong vsize;
}
format
%srcInfo%
Created volume %volumeName% with size %vsize%.
END

/* Start of third template */
facility "VolMgr"; event_type 0x789;
attributes {
      struct sourceInfo srcInfo;
      string diskName;  string diskGroup;
}
format
%srcInfo%
Added disk %diskName% to disk group %diskGroup%.
.fi
.PP

The command
.nf
    evltc VolMgr.t
.fi
will create three files: \fIsourceInfo.to\fR,
\fI0x456.to\fR, and \fI0x789.to\fR in the same
directory as \fIVolMgr.t\fR.


\fBExample 2.\fR
.nf

Given the following formatting template, stored
in the file SCSIadap.t:
 
/* HEADER SECTION */
   facility "LOCAL1";
   event_type 0x3115; 
/* CONST-ATTRIBUTES SECTION */
   const {
         string  repair_action = "Replace SCSI adapter";
   }

/* RECORD-ATTRIBUTES SECTION */
   attributes {
        char    unit_ser_no[8]        "%s";
        ushort  lun                   "%u";
        char    sense_bytes[12]       "%t";
   /* For the next attribute, the various segments of 
    * the format string are automatically concatenated
    * into a single string.
    */
        uchar  recovery_stat  "%b/0x40/INTERFACE_WAS_RESET/"
                              "0x20/RECOVERY_ACTION_STARTED/"
                              "0x10/RECOVERY_ACTION_FAILED/";
   /* The final attribute specification says to display the
    * rest of the bytes in dump format.
    */
   char            extra_data[_R_]     "%t";
   }
   
/* FORMATTING SECTION */
   format
   SCSI interface error: Adapter Ser Num/LUN = %unit_ser_no%/%lun%
            
   \tRecovery Status: %recovery_stat%
   \tSense Bytes:
   %sense_bytes%
                 
   \tRecommended repair action:
   \t\t%repair_action%                 
            
   %extra_data%
 
 
The command
       evltc SCSIadap.t -f
will create the file 0x3115.to in the same directory 
as \fISCSIadap.t\fR and the following will be displayed 
(on stdout):
       
       evl_log_write(LOG_LOCAL1, 12565 /* 0x3115 */, severity, flags,
                  "char[]", 8, unit_ser_no,
                  "ushort", lun,
                  "char[]", 12, sense_bytes,
                  "uchar", recovery_stat,
                  "char[]", _R_, extra_data,
                  "endofdata");
            
       


\fBExample 3.\fR
.nf

These examples illustrate the importing of templates.

The following sequence of statements is legal:

struct statusRegister;
  [definition of struct statusRegister's attributes and format]
END
import chipSet.fpu.statusRegister; /* LEGAL [unless chipSet/fpu
                                   specifies the current directory] */
import chipSet.fpu.dataRegisters;
...
attributes {
  struct statusRegister cpuStatus;      /* refers to struct
                                        defined in this file */
  struct chipSet.fpu.statusRegister fpuStatus;
  struct dataRegisters dataRegs;
 }

However, the following example has an illegal sequence of statements:

struct statusRegister;
  [definition of struct statusRegister's attributes and format]
END
import statusRegister;   /* ILLEGAL */



.fi
.SH "FILES"
.PP
\fI/var/evlog/templates\fR -  Formatting
template repository


CHANGES
Enterprise-Class Event Logging for Linux
========================================

v1.0.0   9/20/2001
Initial  Version


v1.0.1  10/01/2001
- Replaced calls to spin_lock_irq() and spin_unlock_irq() with calls
  to  spin_lock_irqsave() and spin_unlock_irqrestore() in kernel log
  write functions.  This resolved a hang condition experienced with 
  a SCSI device driver.
- Replaced use of sockaddr to sockaddr_un to provide longer socket 
  path name.
- Several minor changes needed for s390 (32 and 64 bit).
- runtests.sh scripts modifed to take option of s390 (32-bit) or
  s390x (64-bit) to specify s390/IBM Z-Series.
- Processing of white-spaces was fixed to correctly process >1 white 
  space.
NOTE:  The patch for GNU C Library, evlog-1.0.1_glibc2.2.3.patch,
is unchanged from version 1.0.0.


v1.1.0  11/09/2001

New features and enhancements in v1.1.0
    
1. Kernel patches for kernel versions 2.4.2 through 2.4.14 will 
   now be posted on the website.  
2. A new library that "wraps" around GNU C Lib and provides
   forwarding of syslog messages to the POSIX event log without
   having to patch and rebuild glibc.
3. Formatting Template support and evltc command (for compiling
   formatting template). Some features are not yet available
   (section numbers in the spec are indicated):
* (6.1.2.2.1.2: Imports and Typedefs) If an import statement specifies
   a struct that has already been created or imported, the import
   statement will be silently ignored.
* (6.1.2.2.1.4: Attributes Section [overall format]) The issue regarding
   const-only structs is still unresolved.  It is currently impossible 
   to define a struct that has only const attributes.
* (6.1.2.2.1.4.3: Attribute Dimension or Bit-Field Width) Bit fields 
   and arrays of structs are not yet supported.
* (6.1.2.2.1.4.5 Attribute Format)   The following is not yet supported:
* Validation of printf-style formats is incomplete.  In particular, you
  are permitted to specify formats that are completely inappropriate for
  the given data type -- e.g. "%f" for a string.
* The %t (dump) format is currently supported only for arrays.
* %b (bitmap) and %v (named-value) specifiers are not yet supported
  for arrays.
* As previously mentioned, arrays of structs are not yet supported;
  so the %Z format specifier is not, either.
* (6.2.3.3 Formatting Templates [API]) The following is not supported:
   * The evl_populatetemplate() function does not currently create the 
     _EXTRA_DATA_ pseudo-attribute
   * The evltemplate_getatt() function does not currently handle names 
     such as "x.y" to reference members of struct-type attributes.

4. Displaying of events logged in the kernel (with the new log write
   functions) on the system console.  The --output option was added to 
   the evlconfig command to turn console logging on/off, or set the
   severity level for displaying to the console.
5. Forwarding of printk() messages to the kernel event buffer must now
   be explicitly enabled when configuring the kernel (forwarding is
   disabled by default).
6. evlfacility command for managing the facility registry.
7. Restricted Logging support. 
8. Added "address" as a data type for formatting templates.
9. Added 2 new purpose arguments for posix_log_query_create() as
   extensions to the draft POSIX standard:  EVL_PRPS_TEMPLATE and
   EVL_PRPS_RESTRICTED.
10. Added EVL_KERNEL_EVENT flag for log_flags member in 
    posix_log_entry.
11. Added varargs versions of log write functions in kernel and 
    user-space.
12. Added support for --new and --timeout options in evlview
    command.
13. Added --templates and --notemplates options to evlview command.
14. Added unlink(PidFile) function in evlogd.c, evlnotifyd.c,
    evlactiond.c.
15. Moved evlogd, evlnotifyd, and evlactiond pid files from /var/evlog
    to /var/run.

Bug Fixes in v1.1.0

1. Removed evlog *.pid files on graceful shutdown or on startup if 
   previous shutdown was non-graceful to fix evlogd daemon startup
   failure.
2. In event notification, fixed a bug where error returns from
   rt_sigqueueinfo() were causing the notification to be erroneously
   disabled.
3. In do_syslog(), now use options 20 and 21 for read/write to kernel
   event buffer, instead of 9 and 10, to avoid conflict with other
   newly added options in do_syslog(). 
4. posix_log_notify_add() is now signal-safe.


v1.1.1  11/28/2001

Bug fixes in v1.1.1

1. (Fix for Bug # 485676) Code added in v1.1.0 for displaying events on the
   system console is incompatible with the re-write of code for displaying of
   messages from printk() first appearing in kernel v2.4.10. This causes the
   kernel to hang or PANIC under certain conditions.  This only affects the
   patches for kernel versions 2.4.10 through 2.4.14. 
2. Added check in the kernel version of posix_log_write() for null-terminated
   string (when log_format is POSIX_LOG_STRING).  Returns EBADMSG if string
   is not null-terminated.  A few other changes to make this function   
   consistent between kernel and user space.
3. Fixed bug causing programs linked with libevl.so to fail at link time.
4. Added missing function call in libevlsyslog.so so that syslog messages
   actually get logged in event log.
5. Fixed several bugs causing problems on s390 and/or 64-bit machines:
   * evltc command reporting bogus errors or causing segfaults
   * Backward seeks through the event log (as used by evlview --tail) failed. 
   * Various memory corruption problems 
6. Updated all header files for C++ compatability.


v1.2.0  02/18/2002

New Features and Enhancements in v1.2.0

1. Events logged in user-space are now passed directly to the evlogd 
   daemon,  instead of being written to the kernel event buffer.
2. The libevl event logging library no longer depends on the event 
   logging kernel patch, meaning that event logging in user-space 	
   does not require  changes in the kernel.  The kernel patch is 

user/cmd/evltc/evltc.c
/*
 * Linux Event Logging for the Enterprise
 * Copyright (c) International Business Machines Corp., 2002
 *
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 *  Please send e-mail to lkessler@users.sourceforge.net if you have
 *  questions or comments.
 *
 *  Project Website:  http://evlog.sourceforge.net/
 *
 */

#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <unistd.h>
#include <string.h>
#include <malloc.h>
#include <errno.h>
#include <locale.h>
#include <assert.h>
#include <sys/wait.h>

#define _GNU_SOURCE
#include <getopt.h>

#include "evl_list.h"
#include "evl_template.h"

extern void printFuncCall(evltemplate_t *t);
static char *cppSrcToTmpFile(const char *fileName, evl_list_t *cppOpts);

static char *progName;
static char *tmpFileName = NULL;
#define MAXTEMPLATES 1000

static void
usage()
{
	fprintf(stderr,
"Usage: %s [-f | --func] [-n | --noto] [-c | --cpp] [cpp opts] srcfile\n"
"  or   %s binfile.to\n",
		progName, progName);
	exit(1);
}

/* Called at exit. */
static void
rmTmpFile()
{
	if (tmpFileName) {
		(void) remove(tmpFileName);
	}
}

main(int argc, char **argv)
{
	char *fileName;
	int status = 0;
	int writeTmpls = 1;
	int printFuncs = 0;
	int runCpp = 0;
	template_t *templates[MAXTEMPLATES];
	size_t nTemplates;
	evl_list_t *cppOpts = NULL;

	static struct option longOptions[] = {
		{"cpp", 0, 0, 'c'},
		{"func", 0, 0, 'f'},
		{"noto", 0, 0, 'n'},
		{0, 0, 0, 0}
	};
	int c;

	progName = argv[0];

	/*
	 * Enable translation of multibyte characters in the "native"
	 * environment.
	 */
	(void) setlocale(LC_CTYPE, "");

	for (;;) {
		c = getopt_long(argc, argv, "cfnD:I:U:", longOptions, NULL);
		if (c == -1) {
			break;
		}
		switch (c) {
		case 'c':
			runCpp = 1;
			break;
		case 'f':
			printFuncs = 1;
			break;
		case 'n':
			writeTmpls = 0;
			break;
		case 'D':
		case 'I':
		case 'U':
		    {
			char *cppOpt;
			runCpp = 1;
			cppOpt = (char*) malloc(strlen(optarg) + 3);
			assert(cppOpt != NULL);
			snprintf(cppOpt, strlen(optarg) + 3, "-%c%s", c, optarg);
			cppOpts = _evlAppendToList(cppOpts, cppOpt);
			break;
		    }
		default:
			usage();
		}
	}

	if (optind != argc-1) {
		usage();
	}
	fileName = argv[optind];
	if (_evlEndsWith(fileName, ".to")) {
		/*
		 * fileName should be a binary template file.
		 * Ignore -f and -n.  Read in the template and print the
		 * evl_log_write() call that could create that record.
		 */
		evltemplate_t *t;

		if (runCpp) {
			fprintf(stderr,
"%s: Cannot specify cpp options with binary template file %s\n", progName, fileName);
			exit(1);
		}
		t = _evlReadTemplate(fileName);
		if (t == NULL) {
			fprintf(stderr,
"%s: Cannot read template from %s\n", progName, fileName);
			exit(2);
		}
		if (t->tm_header.th_type != TMPL_TH_EVLOG) {
			fprintf(stderr,
"%s: Nothing to do with struct template in %s\n", progName, fileName);
			exit(2);
		}
		templates[0] = t;
		nTemplates = 1;
		printFuncs = 1;
	} else {
		/* fileName should be a template source file. */
		char *srcFileName = fileName;
		if (runCpp) {
			tmpFileName = cppSrcToTmpFile(fileName, cppOpts);
			if (!tmpFileName) {
				exit(1);
			}
			srcFileName = tmpFileName;
			(void) atexit(rmTmpFile);
		}
		status = evl_parsetemplates(srcFileName, templates,
			MAXTEMPLATES, &nTemplates, stderr, progName);
		if (status == EBADMSG) {
			/* One or more errors have been reported to stderr. */
			exit(1);
		} else if (status == EINVAL) {
			/*
			 * Probably couldn't open the source file.  errno
			 * should still contain the reason.
			 */
			perror(srcFileName);
			exit(1);
		} else if (status != 0) {
			errno = status;
			perror("evl_parsetemplates");
			exit(2);
		}

		if (writeTmpls) {
			char *dir = _evlGetParentDir(fileName);
			status = evl_writetemplates(dir,
				(const template_t**) templates, nTemplates);
			if (status != 0) {
				errno = status;
				perror("evl_writetemplates");
				exit(1);
			}
		}
	}
	
	if (printFuncs) {
		int i;
		for (i = 0; i < nTemplates; i++) {
			printFuncCall(templates[i]);
		}
	}

	/* Kludge to test evl_freetemplate() */
	if (getenv("FREETMPLS") != NULL) {
		/* Free in reverse order, in case template n uses n-1. */
		int i;
		for (i = nTemplates-1; i >= 0; i--) {
			status = evl_freetemplate(templates[i]);
			if (status != 0) {
				fprintf(stderr,
"evl_freetemplate() of %s failed.\n", templates[i]->tm_name);
			}
		}
	}
	exit (0);
}

/*
 * Run the indicated file through cpp, creating a temp file.  If all goes
 * well, return the name of the temp file; else return NULL.
 */
static char *
cppSrcToTmpFile(const char *fileName, evl_list_t *cppOpts)
{
#ifndef CPPPATH
#define CPPPATH "/lib/cpp"
#endif
	char *tempFile, *tempDir;
	char *result;
	int argc, argsz;
	char **argv;
	pid_t pid;
	int kidStatus;
	evl_listnode_t *head, *end, *p;

	/*
	 * Make the temp file in the same directory as the real source file,
	 * since that's where we should look first for imports.
	 */
	tempDir = _evlGetParentDir(fileName);
	tempFile = (char*) malloc(strlen(tempDir) + 30);
	snprintf(tempFile, strlen(tempDir) + 30, "%s/tmplSrc%d", tempDir, getpid());
	free(tempDir);

	/* /usr/bin/cpp [cpp options] fileName tempFile NULL */
	argsz = 1 + _evlGetListSize(cppOpts) + 1 + 1 + 1;
	argv = (char**) malloc(argsz * sizeof(char*));
	assert(argv != NULL);

	argc = 0;
	argv[argc++] = CPPPATH;
	for (head=cppOpts, end=NULL, p=head; p!=end; end=head, p=p->li_next) {
		argv[argc++] = (char*) p->li_data;
	}
	argv[argc++] = (char*) fileName;
	argv[argc++] = tempFile;
	argv[argc] = NULL;

	pid = fork();
	if (pid == 0) {
		/* I'm the child. */
		(void) execv(CPPPATH, argv);
		perror("Cannot exec " CPPPATH);
		exit(3);
	}

	/* I'm the parent.  Wait for the child to finish. */
	if (pid == -1) {
		perror("fork for cpp");
		exit(3);
	}
	(void) waitpid(pid, &kidStatus, 0);
	if (WIFEXITED(kidStatus) && WEXITSTATUS(kidStatus) == 0) {
		result = tempFile;
	} else {
		result = NULL;
		(void) remove(tempFile);
	}
	free(argv);
	return result;
}
